### PR TITLE
Use Woorelease for deployments

### LINKED ISSUE #1946: Use Woorelease for deployments
To align with our other projects we want to use consistent release tooling. Currently this plugin uses [Sake](https://github.com/skyverge/sake).

I'm proposing we use Sake to build the plugin ZIP file (`npx sake zip`) and use Woorelease to actually ship the release. [Woorelease](https://github.com/woocommerce/woorelease) actually doesn't handle building ZIP files so we can't use it for that anyway.

### PR DESCRIPTION
**Pull request by danielbitzer**
### Changes proposed in this Pull Request:

Makes it possible to use Woorelease to do a deployment. Sake still generates the ZIP file but is responsible for far less.

Closes #1946 .

### Things to check
- [x] `.pot` file is updated
- [ ] Version is updated (Sake normally does this) `changelog.txt`, `readme.txt`, plugin header, plugin constant
- [x] JS assets are generated
- [x] unwanted files/directories are not included
- [x] smoke test build, no js errors, flows work

### How to test the changes in this Pull Request:

1. [Setup Woorelease](https://github.com/woocommerce/woorelease/wiki/Setup)
1. Run `php woorelease.phar simulate --product_version=2.5.0 https://github.com/woocommerce/facebook-for-woocommerce/tree/feature/woorelease-support`

### Changelog entry

> Dev - Change deployment tooling

### CODE REVIEW

**Comment by haszari** [comment]
Tentatively added to 2.5.0. I'll do some testing/review today to confirm this is not going to cause any issues - would love to make the switch and reduce the dependency on Sake.

**Review by haszari — COMMENTED** [review]
Still testing this, overall looks good. Added some inline comments, mostly minor/non blocking.

FYI here's my plan for testing:

- Ensure WooRelease is installed & configured on my machine & on latest version `2.0.3` (I already have it set up).
- Simulate a release and see what happens.
- Is the built zip shipshape:  
  - Changes are in .pot file.
  - Version numbers updated in various places.
  - JS & front end assets are built and contain recent changes.
  - Might commit some fake code changes to a branch (off this branch) to test end to end.
- Try generating the changelog - this is a useful feature.
- Smoke test the built zip - make sure the key flows are working and nothing's obviously broken.

Then if that works out will proceed from there 🤞

File: .nvmrc [hunk 1]
```diff
@@ -0,0 +1 @@
+v12
```
[the remark this answers is not in the document]
Comment (danielbitzer) [reply .nvmrc:1]: Yep its an nvm config file. I'm running node 15.11.0 and I can't build most things like WooCommerce, AutomateWoo or this plugin without using switching down to v12. Gan recommended using nvm but I don't know what's recommended widely in Woo.

File: bin/build-zip.sh [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+echo "Building @wordpress/scripts assets..."
+npm run build:assets
+
+# Sake installs composer and makes a ZIP
+echo "Building zip with Sake..."
+npx sake zip || exit
+
+echo "Moving zip..."
+find $BUILD_PATH -name "facebook-for-woocommerce.*.zip" -exec cp {} $FINAL_ZIP \; || exit;
```
[the remark this answers is not in the document]
Comment (danielbitzer) [reply bin/build-zip.sh:24]: True, its looks like we can just use the simpler `cp` command.

> What would happen if there are multiple version zips in ${PROJECT_PATH}/build - are we assuming that there's only one?

[Sake always cleans out any ZIP files](https://github.com/skyverge/sake/blob/master/tasks/clean.js#L15) before running but we could always delete `build/` in our script just to be sure.

File: bin/build-zip.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
+
+PLUGIN_SLUG="facebook-for-woocommerce"
+PROJECT_PATH=$(pwd)
+BUILD_PATH="${PROJECT_PATH}/build"
+FINAL_ZIP="${PROJECT_PATH}/${PLUGIN_SLUG}.zip"
+
+echo $PROJECT_PATH
```
[the remark this answers is not in the document]
Comment (danielbitzer) [reply bin/build-zip.sh]: I've tweaked this and a few other things you've raised in cb7aec0b0c9cdeb3964db9392f33f007c559f256

**Comment by haszari** [comment]
Merged latest master to get latest changes and #1955 in particular :)

File: .nvmrc (hunk 1 above)
[the remark this answers is not in the document]
Comment (rcstr) [reply .nvmrc:1]: I like the idea of specifying a node version that would work without errors. I'm not sure if `.nvmrc` it's the right place

1. [`.nvmrc` doesn't enforce a `node` version, you have to switch to the version using `nvm use`](https://github.com/nvm-sh/nvm#nvmrc).
2. `engines` in `package.json` won't enforce either, [but at least will display a warning in the terminal](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#engines).

set `engines` seems like a more standard approach imo

**Review by rcstr — APPROVED** [review]
I created [`release/test-rommel`](https://github.com/woocommerce/facebook-for-woocommerce/tree/release/test-rommel) branch to test this, on this branch I:

- Updated a few translated strings
- Added a few more _version_ constants followed by `//WRCS: DEFINED_VERSION.` comment
- Ran `wr simulate --product_version=2.7.0 https://github.com/woocommerce/facebook-for-woocommerce/tree/release/test-rommel` where `wr` it's a alias for `woorelease`
- `.pot` files look good, contained changes I did
- **Versioning works on `readme.txt` and `changelog.txt` but didn't see any change on PHP constants in build/zip**.
- https://github.com/woocommerce/facebook-for-woocommerce/blob/release/test-rommel/class-wc-facebookcommerce.php#L25-L26
- Assets were generated
- No ignored files, folders present in zip or build
[facebook-for-woocommerce.2.7.0.zip](https://github.com/woocommerce/facebook-for-woocommerce/files/6475279/facebook-for-woocommerce.2.7.0.zip)

For me, the Version in code constants wasn't updated to `2.7.0`. @danielbitzer is this expected? maybe I'm doing something wrong

**Comment by haszari** [comment]
> For me, the Version in code constants wasn't updated to 2.7.0. @danielbitzer is this expected? maybe I'm doing something wrong.

I think we'll still need to update version numbers and push as before, then run build/release. More support for version numbers maybe coming in next WooRelease.

File: .nvmrc (hunk 1 above)
[the remark this answers is not in the document]
Comment (haszari) [reply .nvmrc:1]: > I'm running node 15.11.0 and I can't build most things like WooCommerce, AutomateWoo or this plugin without using switching down to v12.

Side note: I believe this (and other plugins) should build fine with node LTS - currently 14.17. I was on 14.16 until updated this morning. (I stick with LTS, using `nvm --use lts` and `nvm --install lts`.

[the remark this answers is not in the document]
Comment (haszari) [reply .nvmrc:1]: Here's the `engines` that [Woo core is using](https://github.com/woocommerce/woocommerce/blob/trunk/package.json#L98):

```json
  "engines": {
    "node": "^12.20.1",
    "npm": "^6.14.6"
  },
```

I'll log an issue for this, I don't think it blocks this PR but would be good to add.

[the remark this answers is not in the document]
Comment (haszari) [reply .nvmrc:1]: #1957

**Comment by haszari** [comment]
Note also I see these warnings when I simulate (forgot to include this in review):

```
WR [01:31:36] [WARNING] No constant version define for facebook-for-woocommerce, skipping...
WR [01:31:36] [WARNING] No version declared in package.json for facebook-for-woocommerce. Skipping...
WR [01:31:36] [WARNING] No version declared in package-lock.json for facebook-for-woocommerce. Skipping...
```

**Comment by danielbitzer** [comment]
Thanks @haszari and @rcstr for the testing and feedback!

I feel fine about moving the version constant into the main plugin folder so Woorelease can automatically update it. I've made this change in 422d695514b2dda86a2e8a4f74641606a179ba4f.

I've also removed the .nvmrc file addition in this PR, we should probably address node versions in a separate PR.

**Review by haszari — APPROVED** [review]
Retested and we can work with this. There are still things we can improve but we can do that as follow up. 

Here's my proposed workflow/checklist for releasing:

- Make a release branch.
- Update version numbers and changelog (manually).
  - Version number needs changing in these files: 
  - `readme.txt`
  - `changelog.txt`
  - `package.json` & `package-lock.json`
  - `facebook-for-woocommerce.php`
  - (`class-wc-facebookcommerce.php` not needed anymore)
- Code freeze/RC day – build and simulate to generate a prerelease / RC:
  - `woorelease simulate --product_version=2.5.0 https://github.com/woocommerce/facebook-for-woocommerce/tree/rua/try/woorelease-build`
  - Grab zip from tmp folder and upload to GitHub prerelease RC.
- Update changelog & release branch if any fixes / PRs merged during RC period.
- Release day – release for real:
  - `woorelease release --product_version=2.5.0 https://github.com/woocommerce/facebook-for-woocommerce/tree/rua/try/woorelease-build`

**Pull request merged**